sample_code/clean_unknown_vertices/ExprFunctions.hpp

```diff
@@ -541,6 +541,93 @@ inline string CombineTwoNumbers(int64_t one, int64_t two) {
+      gshared_ptr<topology4::GraphUpdates> graphupdate,
+      uint32_t eid,
+      topology4::EdgeAttribute* eAttr,
+      int64_t src,
+      int64_t tgt) {
+    auto attr_up = graphupdate->GetEdgeAttributeUpdate(eid);
+    if (attr_up->Set(eAttr)) {
+      graphupdate->UpsertEdge(
+          topology4::DeltaVertexId(context.GraphAPI()->GetVertexType(src), src),
+          topology4::DeltaVertexId(context.GraphAPI()->GetVertexType(tgt), tgt),
+          attr_up,
+          eid);
+    } else {
+      GEngineInfo(InfoLvl::Brief, "RecoverOneEdge")
+      << "[ERROR]: Insert edge (src: "
+      << src
+      << ", tgt: "
+      << tgt
+      << ", eid: "
+      << eid
+      << ") failed.";
+    }
+  }
+
+inline void RecoverEdges(ServiceAPI* serviceapi,
+      EngineServiceRequest& request,
+      gpr::Context& context,
+      const std::string& edgeType,
+      const std::string& rEdgeType,
+      VERTEX src,
+      ListAccum<int64_t>& fEdgeList,
+      ListAccum<int64_t>& rEdgeList,
+      int64_t& missingForward,
+      int64_t& missingReverse,
+      bool dryrun) {
+
+    auto fEid = serviceapi->GetTopologyMeta()
+      ->GetEdgeTypeId(edgeType, request.graph_id_);
+    auto rEid = serviceapi->GetTopologyMeta()
+      ->GetEdgeTypeId(rEdgeType, request.graph_id_);
+
+    auto& fTgtList = fEdgeList.data_;
+    auto& rTgtList = rEdgeList.data_;
+    std::sort(rTgtList.begin(), rTgtList.end());
+
+    //scan the two sorted list to figure out which target has missing edge
+    uint32_t i = 0,  j = 0;
+    auto graphupdate = serviceapi->CreateGraphUpdates(&request);
+    while(i < fTgtList.size() && j < rTgtList.size()) {
+      if (fTgtList[i] == rTgtList[j]) {
+        ++i;
+        ++j;
+      } else if (fTgtList[i] > rTgtList[j]) {
+        //rEdgeList[j] is one target that only have reversed edge
+        auto eAttr = context.GraphAPI()->GetOneEdge(rTgtList[j], src.vid, rEid);
+        RecoverOneEdge(context, graphupdate, rEid, eAttr, rTgtList[j], src.vid);
+        ++j;
+        ++missingReverse;
+      } else {
+        //fEdgeList[i] is one target that only have reversed edge
+        auto eAttr = context.GraphAPI()->GetOneEdge(src.vid, fTgtList[i], fEid);
+        RecoverOneEdge(context, graphupdate, fEid, eAttr, src.vid, fTgtList[i]);
+        ++i;
+        ++missingForward;
+      }
+    }
+    while(i < fTgtList.size()) {
+      auto eAttr = context.GraphAPI()->GetOneEdge(src.vid, fTgtList[i], fEid);
+      RecoverOneEdge(context, graphupdate, fEid, eAttr, src.vid, fTgtList[i]);
+      ++i;
+      ++missingForward;
+    }
+    while(j < rTgtList.size()) {
+      //rEdgeList[j] is one target that only have reversed edge
+      auto eAttr = context.GraphAPI()->GetOneEdge(rTgtList[j], src.vid, rEid);
+      RecoverOneEdge(context, graphupdate, rEid, eAttr, rTgtList[j], src.vid);
+      ++j;
+      ++missingReverse;
+    }
+    //clean up the target list to release memory
+    fEdgeList.clear();
+    rEdgeList.clear();
```

Comment:
Has it already been cleared in the gsql?

Reply:
yes, double cleared. Will remove one